### PR TITLE
Update inference scaling to support multi-GPU functions

### PR DESCRIPTION
**Pull request by billschereriii**
Update inference scaling to use multi-GPU support functions. On a machine with 8 GPUs, we obtained a 4x performance improvement with this technique.

### CODE REVIEW

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 1]
```diff
@@ -125,20 +130,21 @@ void run_mnist(const std::string& model_name,
...
 
   MPI_Barrier(MPI_COMM_WORLD);
-  if(!rank)
+  if (!rank)
     std::cout<<"All ranks have Resnet image"<<std::endl;
 
   double loop_start = MPI_Wtime();
-  for (int i=0; i<100; i++) {
-
+  for (int i = 0; i < 101; i++) {
```
Comment (Spartee) [inline cpp-inference/inference_scaling_imagenet.cpp:147]: any reason for changing this number? might be nice to have this be a env var or cli arg

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 2]
```diff
@@ -152,32 +158,37 @@ void run_mnist(const std::string& model_name,
...
+    unpack_tensor_times.push_back(delta_t);
 
   }
   double loop_end = MPI_Wtime();
   delta_t = loop_end - loop_start;
 
   // write times to file
-  for (int i=0; i<100; i++) {
-
+  for (int i = 1; i < 101; i++) { // Skip first run as it's warmup
```
Comment (Spartee) [inline cpp-inference/inference_scaling_imagenet.cpp:191]: ok I see the reason now. would it be better to invoke this once outside the main loop? I feel like we could re-use this to do it if we made this part into a function with the number of iterations as an argument. either way I understand this now so disregard previous question

File: driver.py [hunk 3]
```diff
@@ -602,7 +618,7 @@ def write_run_config(path, **kwargs):
         "name": name,
         "path": path,
         "smartsim_version": smartsim.__version__,
-        "smartredis_version": "0.3.0", # TODO put in smartredis __version__
+        "smartredis_version": "0.4.0", # TODO put in smartredis __version__
```
Comment (Spartee) [inline driver.py:621]: this should be ``smartredis.__version__`` but we still need to put that into the python client library in the top level __init__.py. Either way, the next version of SmartRedis will be 0.3.1 if thats what you were going for here.

**Review by Spartee — APPROVED** [review]
Awesome! Approved with the exception of 1 fix needed around the SmartRedis version. We will want to take a similar approach to what we did with SmartSim once the SmartRedis ``__version__`` is specified in the client library.

File: cpp-inference/inference_scaling_imagenet.cpp (hunk 1 above)
Comment (billschereriii) [reply cpp-inference/inference_scaling_imagenet.cpp:147]: I bumped it by one to account for the warmup run we aren't including in the results

File: cpp-inference/inference_scaling_imagenet.cpp (hunk 2 above)
Comment (billschereriii) [reply cpp-inference/inference_scaling_imagenet.cpp:191]: We probably should add config settings for number of warmup runs and number of test runs, though that's probably outside the scope of this change.

As far as moving it to a separate function, I worry about code duplication.

File: driver.py (hunk 3 above)
Comment (billschereriii) [reply driver.py:621]: I'll tweak this to 0.3.1 then as a stopgap measure

**Review by Spartee — APPROVED** [review]
LGTM

**Review by al-rigazzi — COMMENTED** [review]
If we are going with the 101 iterations, then we should just clearly indicate somewhere (in README.md and maybe in docstrings) that the first one is not taken into account for metrics. Just to make sure it does not look like we are cheating.

**Comment by billschereriii** [comment]
Good call, Al, I'll make sure to document that we have a warmup cycle

**Pull request merged**